spec.html

```diff
@@ -8794,6 +8834,24 @@ <h1>CreateBuiltinFunction ( _steps_, _internalSlotsList_ [ , _realm_ [ , _protot
       </emu-alg>
       <p>Each built-in function defined in this specification is created by calling the CreateBuiltinFunction abstract operation.</p>
     </emu-clause>
+
+    <emu-clause id="sec-createbuiltingeneratorfunction" aoid="CreateBuiltinGeneratorFunction">
+      <h1>CreateBuiltinGeneratorFunction ( _steps_ [ , _iteratorKind_ [ , _generatorPrototype_ [ , _generatorInternalSlotsList_ ]]]) </h1>
+      <p>The abstract operation CreateBuiltinGeneratorFunction create a <dfn id="sec-builtin-generator-function">built-in generator function</dfn> object. The generator object it returns can have extra internal slots.</p>
+      <p>The abstract operation CreateBuiltinGeneratorFunction takes arguments _steps_, and optional arguments _generatorInternalSlotsList_ (a List of internal slot names that the returned generator has), _generatorPrototype_ (the prototype of the returned generator, should inherit from %IteratorPrototype%). It performs the following steps when called:</p>
+      <emu-alg>
+        1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification. It can contain Yield abstract operations to yield an IteratorResult object.
+        1. If _generatorPrototype_ is not present, set _generatorPrototype_ to %Iterator.prototype%.
+        1. Append [[GeneratorState]], [[GeneratorContext]], [[IteratorKind]] to _generatorInternalSlotsList_.
+        1. Let _wrappedSteps_ be:
+          1. Let _generator_ be ? OrdinaryObjectCreate(_generatorPrototype_, _generatorInternalSlotsList_).
+          1. Set _generator_.[[IteratorKind]] to _iteratorKind_.
+          1. Perform ? GeneratorStart(_generator_, _steps_).
```

Comment:
Is there an existing example of handling steps and <kbd>FunctionBody</kbd> I can refer to?

Reply:
Not in the main spec, no. There might be in some proposal, but I doubt it.

But it might be that all you need is:
```
If _generatorBody_ is a Parse Node, then
    Let _result_ be the result of evaluating _generatorBody_.
Else,
    Assert: _generatorBody_ is either a set of algorithm steps [etc]
    Let _result_ be the result of performing the action described by _steps_.
```
This would be better if 'steps' is replaced with an abstract closure, as suggested elsewhere.

(Alternatively, you could change `GeneratorStart` so that it always takes a closure, and then, at the call-point that currently passes `|FunctionBody|`, construct a closure to evaluate `|FunctionBody|`, and pass that.)

Reply:
Does the abstract closure can be "suspended"?

Reply:
I don't see why not, though I guess it hasn't been established yet.

Reply:
> ```
> Let _result_ be the result of evaluating _generatorBody_.
> ```

Does it also work if generatorBody is an abstract closure?

Reply:
No, abstract closures "are invoked using function application style such as _closure_(_arg1_, _arg2_)".